### DISCUSSION TITLE
Multi-phase monitoring PZEM004T - Per phase Today/Yesterday/Total Energy

### DISCUSSION THREAD

**Discussion by Montreal666** [arendst/Tasmota#13542] [Q&A] [answered]
In order to optimize components, I'd like to use the Multi-phase/adresses feature to monitor 3 different single phase circuits/devices.

IE: one Sonoff basic + 3x PZEM004T ( setup as per https://github.com/arendst/Tasmota/issues/2315 )

Therefore, I'd like to get Today/Yesterday/Total Energy separately for each "phase".
Would it be possible to add a "setoption" or something equivalent to show theses 3 items in the same "xxxxx/xxxxx/xxxxx" format as the the remaining data (voltage, current, power, etc.) As I would use the "yesterday" output to track daily energy draw for each PZEM/"phase".

![image](https://user-images.githubusercontent.com/56667580/139783699-5dc91b17-3ac6-4456-8d07-d78ea933c5e0.png)

I could add some calculations to generate the numbers in my home automation platform or in InfluxDB/Grafana, but it would great to have the option to get the data straight from Tasmota.

Let me know if this is feasible ,

Thanks again for this great project!

Cheers,

**Comment by ascillato** [comment, marked as answer]
This is your lucky day. That is already supported.

Docs: https://tasmota.github.io/docs/Commands/#setoption129

- Command ``SetOption129 1`` to enable split total energy results
- Commands ``EnergyTotal<phase>``, ``EnergyToday<phase>`` and ``EnergyYesterday<phase>`` to (re)set energy values

----------------

Issues: #13030 , #6160
Commit: https://github.com/arendst/Tasmota/commit/d3fc03d3980c25cd8b66063405bcab12adadaff5

**Reply by Montreal666** [reply]
My lucky day indeed ;) implemented 4 weeks ago....
Couldn't find it as I was looking for PZEM references.
Thanks for the prompt reply, now to testing. Should be a great addition to my energy monitoring set up.
Cheers,

**Reply by onley** [reply]
Along the lines of the original post using the ModuleAddress to access multiple PZEM devices in a 3 phase arrangement, would it be possible to expand the number of modules from 3 to ? for people who want to use many PZEM modules on a single host (SonOff, Wemo...) as single phase circuit monitors with independent mqtt payloads for each module?

**Reply by barbudor** [reply]
Yes and no
The Tasmota Energy model is meant to manage up to 3 phases. Thus the inherent limitation to 3.

In theory it would be possible to make a different driver for PZEM not following the Tasmota Energy model but the more generic Sensor model. A sensor which would report Energy instead of reporting Temperature, Humidity etc....
In that case, everything about EnergyYesterday/Today/Total would not be applicable, only reporting.
This is what Theo chose for the Sonoff SPM (xdrv_86_esp32_sonoff_spm.ino) whcih can support up to 28 relays with power monitoring (28 being another hard limit in Tasmota for relays).

**Reply by onley** [reply]
So it sounds like no unless I write a different driver and I'm afraid I'm not capable of that. Thank you for this information.

**Reply by gemu2015** [reply]
if you choose MODBUS compatible PZEM devices you can use the SML driver. it allows for more then 3 MODBUS devices an the same bus

**Reply by onley** [reply]
I built a firmware with the SML driver on gitpod but I do not understand how to configure it. Could you possibly point me to some configuration information for a PZEM-004t V3.0 module?

**Reply by gemu2015** [reply]
read the sml docs, there are some examples of other MODBUS devices that you can adapt.

https://tasmota.github.io/docs/Smart-Meter-Interface/

**Reply by Constipated-Monk** [reply]
what is the sample size for EnergyToday and EnergyYesterday? same as the amount of telemetry messages, fixed size or other?
